src/board/index.ts

```diff
@@ -61,79 +61,131 @@ export class Board {
         "buttonB",
-        onSensorChange
+        this.svg.querySelector("#ButtonB")!,
+        "button B",
+        onChange
       ),
     ];
     this.pins = Array(33);
     this.pins[MICROBIT_HAL_PIN_FACE] = new Pin(
-      this.svg.querySelector("#Logo")!,
       "pinLogo",
-      onSensorChange
+      this.svg.querySelector("#Logo")!,
+      "logo",
+      onChange
     );
-    this.pins[MICROBIT_HAL_PIN_P0] = new Pin(null, "pin0", onSensorChange);
-    this.pins[MICROBIT_HAL_PIN_P1] = new Pin(null, "pin1", onSensorChange);
-    this.pins[MICROBIT_HAL_PIN_P2] = new Pin(null, "pin2", onSensorChange);
+    this.pins[MICROBIT_HAL_PIN_P0] = new Pin("pin0", null, "pin 0", onChange);
+    this.pins[MICROBIT_HAL_PIN_P1] = new Pin("pin1", null, "pin 1", onChange);
+    this.pins[MICROBIT_HAL_PIN_P2] = new Pin("pin2", null, "pin 2", onChange);
     this.audio = new Audio();
     this.temperature = new RangeSensor("temperature", -5, 50, 21, "°C");
-    this.accelerometer = new Accelerometer(onSensorChange);
+    this.accelerometer = new Accelerometer(onChange);
     this.microphone = new Microphone(
       this.svg.querySelector("#LitMicrophone")!,
-      onSensorChange
+      onChange
     );
     this.radio = new Radio(
       this.notifications.onRadioOutput.bind(this.notifications),
+      onChange,
       this.ticksMilliseconds.bind(this)
     );
 
-    this.sensors = [
-      this.display.lightLevel,
-      this.temperature,
-      this.microphone.soundLevel,
-      this.buttons[0].button,
-      this.buttons[1].button,
-      this.pins[MICROBIT_HAL_PIN_FACE].pin,
-      this.pins[MICROBIT_HAL_PIN_P0].pin,
-      this.pins[MICROBIT_HAL_PIN_P1].pin,
-      this.pins[MICROBIT_HAL_PIN_P2].pin,
-      ...this.accelerometer.sensors,
-    ];
-    this.sensorsById = new Map();
-    this.sensors.forEach((sensor) => {
-      this.sensorsById.set(sensor.id, sensor);
-    });
     this.stoppedOverlay = document.querySelector(".play-button-container")!;
     this.playButton = document.querySelector(".play-button")!;
     this.initializePlayButton();
     // We start stopped.
     this.displayStoppedState();
     this.playButton.addEventListener("click", () =>
-      window.parent.postMessage(
-        {
-          kind: "request_flash",
-        },
-        "*"
-      )
+      this.notifications.onRequestFlash()
     );
   }
 
-  getSensor(id: string): Sensor | undefined {
-    return this.sensorsById.get(id);
+  getState(): State {
+    return {
+      radio: this.radio.state,
+      buttonA: this.buttons[0].state,
+      buttonB: this.buttons[1].state,
+      pinLogo: this.pins[MICROBIT_HAL_PIN_FACE].state,
+      pin0: this.pins[MICROBIT_HAL_PIN_P0].state,
+      pin1: this.pins[MICROBIT_HAL_PIN_P1].state,
+      pin2: this.pins[MICROBIT_HAL_PIN_P2].state,
+
+      accelerometerX: this.accelerometer.state.accelerometerX,
+      accelerometerY: this.accelerometer.state.accelerometerY,
+      accelerometerZ: this.accelerometer.state.accelerometerZ,
```

Comment:
This is kinda messy/redundant. See also calls in jshal.js.

Reply:
Living with this for now. It's not API and it does make it easier to implement setValue.